### ISSUE TITLE
Ordering between zfs and resume initcpio hooks

### ISSUE THREAD

**Issue by netvl** [archzfs/archzfs#284]
I use the encrypted ZFS root setup, using native ZFS encryption facilities. The structure of datasets is as follows:
```
NAME                           USED  AVAIL     REFER  MOUNTPOINT
zroot                          392G   507G       96K  none
zroot/encrypted                392G   507G      192K  none
zroot/encrypted/home           368G   507G      192K  none
zroot/encrypted/home/default   368G   507G      368G  /home
zroot/encrypted/root          6,77G   507G      192K  none
zroot/encrypted/root/default  6,77G   507G     6,77G  /
zroot/encrypted/swap          17,0G   524G      108K  -
```
As you can see, I have a swap partition, which is also encrypted.

Swap is successfully mounted when I boot the system in the usual way. However, the way the existing initcpio hooks are implemented makes it impossible to use hibernation in my setup.

From what I see in the script, the `zfs` hook does a `zfs import` when invoked, but it does not use the `-l` option which makes the `import` command to ask for the key of the encrypted dataset. Instead, the actual decryption and mounting is deferred to the `zfs_mount_handler` function, which gets run after all other hooks are run. The `resume` hook, which is necessary for the system to wake up from hibernation, however, expects the swap device to be already available by the point it is run.

This can be observed on a real system as the `resume` hook waiting for 10 seconds for the swap device, and only after this the encryption key is asked.

Frankly, I'm not sure how to fix this correctly. Adding `-l` option to `zfs import` would result into the system asking for passwords for datasets which are potentially unnecessary for the boot process. It would work fine for my simple setup, but for more complex setups I can see it being inconvenient. But the way it is done now means that it is not possible to do hibernation with encrypted swap on ZFS, which I think is a very unfortunate limitation, because it would either force the user to use unencrypted swap, or to manage the swap device in a different way, e.g. with LVM or a separate fixed-size partition with LUKS, which makes it much harder to configure and less convenient to use.

**Comment by michaelherold** [comment]
@netvl, I am curious if you ever got this working? I have a similar setup, though my swap is in an unencrypted dataset. Even though that's the case, I have the same output that you experience:

1. Wait 10 seconds for the mount to be available
2. Get prompted for the encryption password

That doesn't seem right but it could, perhaps, be related to what you're saying with the lack of a `-l` flag on the import. Maybe the import command tries to import every dataset at once?

I haven't dug into the `zfs` hook yet so I'm not sure if that makes any sense.

**Comment by netvl** [comment]
@michaelherold unfortunately I've never got this working.  I don't think I ever tried getting out from hibernation with an unencrypted swap either.

**Comment by stevleibelt** [comment]
Shouldn't this be closed, because it is old/possible outdated?
